File: DFRobot_ASR.cpp
/*!
   @file DFRobot_ASR.h
   @brief DFRobot_ASR 类的基础结构，基础方法的实现
   @copyright   Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
   @licence     The MIT License (MIT)
   @author [fengli](li.feng@dfrobot.com)
   @version  V1.0
   @date  2020-08-17
   @get from https://www.dfrobot.com
   @https://github.com/DFRobot/DFRobot_ASR
*/
#include "DFRobot_ASR.h"
DFRobot_ASR::DFRobot_ASR(TwoWire *pWire, uint8_t address){

  _deviceAddr = address;
  _pWire = pWire;

}
int DFRobot_ASR::begin(eMode_t mode,eMicrophoneMode_t miMode)
{
  _mode = mode;
  _pWire->begin();
  _pWire->beginTransmission(_deviceAddr);
  if(_pWire->endTransmission() != 0){
    DBG("bus data access error");
    return ERR_DATA_BUS;
   }
  writeReg(ASR_BEGIN);
  if(miMode == MIC){
     writeReg(ASR_MIC_MODE);
  } else {
     writeReg(ASR_MONO_MODE);
  }
  
  
  delay(50);
  return ERR_OK;
}
void DFRobot_ASR::start()
{
  writeReg(ASR_START);
  delay(50);
}
bool DFRobot_ASR::addCommand( char *words, char idNum)
{
   uint16_t length = strlen(words);
   uint16_t lenTemp = length;
   //Serial.println(length);
   if(length > 72) return false;
   _pWire->beginTransmission(_deviceAddr);
   _pWire->write(ASR_ADDCOMMAND);
   _pWire->write(idNum);
   _pWire->write(length);
   _pWire->endTransmission();
   
   //_pWire->beginTransmission(_deviceAddr);
   while(length){
      if(length > 28){
         lenTemp =28;
      } else {
         lenTemp = length;
     }
   
      _pWire->beginTransmission(_deviceAddr);
      for(uint8_t i =0;i<lenTemp;i++){
        _pWire->write(words[i]);
        DBG(words[i]);
      }
      if( _pWire->endTransmission() != 0 ) {
          DBG("ERR_DATA_BUS");
          return ERR_DATA_BUS;
      }
      length -= lenTemp;
      words += lenTemp;
   }
   //if( _pWire->endTransmission() != 0 ) {
    //   delay(10);
   //    return false;
  // }
   
   _pWire->beginTransmission(_deviceAddr);
   
   _pWire->write(ASR_ADDCOMMAND_END);
   if( _pWire->endTransmission() != 0 ) {
       delay(10);
       return false;
   }
   delay(20);
   return true;
}

int DFRobot_ASR::read(){

  int result = 0;
  switch (_mode)
  {
     case      BUTTON: readReg(ASR_BUTTON,&result,1);
                       break;   
     case       LOOP: readReg(ASR_LOOP,&result,1);
               break;
     case      PASSWORD: readReg(ASR_PASSWORD,&result,1);
               break;          
     default: break;
  }
  if(_mode == PASSWORD){
    idle++;
    if(idle >= 500){
       writeReg(ASR_IDLE);
       idle = 0;
    }
  }
  if(result == 0xff){
    return -1;
  }
  else {
    idle = 0;
    return result;
  }
}
void DFRobot_ASR::setI2CAddr(uint8_t addr){

  if(addr >127)  addr = 127;
  writeReg(ASR_SET_IIC_ADDR,addr);

}
void DFRobot_ASR::writeReg(uint8_t reg, uint8_t data)
{
  _pWire->beginTransmission(_deviceAddr);
  _pWire->write(reg);
  _pWire->write(data);
  _pWire->endTransmission();
}
uint8_t DFRobot_ASR::writeReg(uint8_t reg)
{

    _pWire->beginTransmission(_deviceAddr);
    _pWire->write(reg);
    if( _pWire->endTransmission() != 0 ) {
        return false;
    }
    return true;
}



uint8_t DFRobot_ASR::readReg(uint8_t reg, void* pBuf, size_t size)
{
  if(pBuf == NULL){
    DBG("pBuf ERROR!! : null pointer");
  }
  uint8_t * _pBuf = (uint8_t *)pBuf;
    _pBuf[0] = 0xff;
    unsigned char i = 0;  
    /* Indicate which register we want to read from */
    if (!writeReg(reg)) {
        return -1;
    }
    delay(18);
   _pWire->requestFrom(_deviceAddr, size);
   while (_pWire->available()) {
        if (i >= size) {
            return -1;
        }
        _pBuf[i] = _pWire->read();
        i++;
    }
    /* Read block data */    
    return i;

}

File: python/raspberrypi/DFRobot_ASR.py
# -*- coding: utf-8 -*
'''!
  @file DFRobot_ASR.py
  @brief The speech recognition module can recognize the phrases pre-written to the sensor and return the corresponding number.
  @copyright   Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
  @license     The MIT license (MIT)
  @author [fengli](li.feng@dfrobot.com)
  @version  V1.0
  @date  2020-08-17
  @https://github.com/DFRobot/DFRobot_ASR
'''
import serial
import time
import smbus
               
class DFRobot_ASR(object):
  ''' register configuration '''
  I2C_ADDR                = 0x50
  ASR_BEGIN               = 0xA1  #module init command
  ASR_ADDCOMMAND          = 0xA2  #command to start writing command
  ASR_ADDCOMMAND_END      = 0xA3  #command to end writing command
  ASR_START               = 0xA4  #command to start recognition
  ASR_LOOP                = 0xA5  #Loop mode command
  ASR_BUTTON              = 0xA7  #Button mode command
  ASR_PASSWORD            = 0xA6  #Password mode command
  ASR_IDLE                = 0xA8  #Idle mode command
  ASR_MIC_MODE            = 0xA9  #mic mode command
  ASR_MONO_MODE           = 0xAA  #mono mode command
  ASR_SET_IIC_ADDR        = 0xAB  #Set address command
  LOOP                    = 0x01  #Loop mode
  PASSWORD                = 0x02  #Password mode
  BUTTON                  = 0x03  #Button mode
  MIC                     = 0x04
  MONO                    = 0x05
  
  ''' Conversion data '''
  #txbuf      = [0]
  #_addr      =  0x50
  #_mode      =  0
  #   idle =    0
  def __init__(self ,bus,address):
    self.i2cbus = smbus.SMBus(bus)
    self._addr = address
    self.idle =    0

  def begin(self ,mode,miMode):
  '''
    @fn begin
    @brief Init function
    @param mode Speech recognition mode
    @param miMode Mic mode
    @return Returning 0 indicates init succeeded, non-zero indicates init failed, return error code.
  '''
    self._mode = mode
    self.write_data(self.ASR_BEGIN)
    if miMode == self.MIC:
       self.write_data(self.ASR_MIC_MODE)
    else:
       self.write_data(self.ASR_MONO_MODE)
    time.sleep(0.05)

  def start(self):
  '''
    @fn start
    @brief The speech recognition module start to recognize.
  '''
    self.write_data(self.ASR_START)
    time.sleep(0.05)

  def addCommand(self,words,idNum):
  '''
    @fn addCommand
    @brief Add phrases to the module
    @param words Character strings of the phrase
    @param idNum The id number of the phrase
    @return Boolean type, the result of seted
    @retval Adding phrases succeeded
    @retval Adding phrases failed
  '''
    #word = str(chr(idNum))+words+str(chr(self.ASR_ADDCOMMAND_END))
    #print(word[0])
    self.write_data(self.ASR_ADDCOMMAND)
    self.write_data(idNum)
    
    word = list(words)
    self.write_data(len(word))
    #print(len(word))
    for i in range(0,len(word)):
      self.write_data(ord(word[i]))
      #print(word[i])
    #self.i2cbus.write_i2c_block_data(self._addr ,idNum ,word)
    self.write_data(self.ASR_ADDCOMMAND_END)
    
  def read(self):
  '''
    @fn read
    @brief Read the recognized phrase
    @return Return the id number of the phrase
  '''
    rslt = -1
    #print(self._mode)
    if self._mode == self.BUTTON:
        rslt = self.read_reg(self.ASR_BUTTON,1)
    elif self._mode == self.LOOP:
        rslt = self.read_reg(self.ASR_LOOP,1)
    elif self._mode == self.PASSWORD:
        rslt = self.read_reg(self.ASR_PASSWORD,1)
    if self._mode == self.PASSWORD:
        self.idle = self.idle + 1
        if self.idle >= 500:
           self.write_data(self.ASR_IDLE)
           self.idle = 0
    if rslt == 0xff:
       return -1
    else:
       self.idle=0
       return rslt


  def setI2CAddr(self ,addr):
  '''
    @fn setI2CAddr
    @brief Set module i2c address (take effect after powering on again)
    @param addr i2c address to be set (0-127)
  '''
    if addr > 127:
        addr = 127
    self.i2cbus.write_byte_data(self._addr,self.ASR_SET_IIC_ADDR ,addr)
    #print(self._addr)
    #print(self.ASR_SET_IIC_ADDR)
    #print(addr)

  def write_data(self, data):
    self.i2cbus.write_byte(self._addr ,data)
    
  def write_reg(self, reg, data):
    self.i2cbus.write_byte(self._addr ,reg)
    self.i2cbus.write_byte(self._addr ,data)

  def read_reg(self, reg ,len):
    self.i2cbus.write_byte(self._addr,reg)
    time.sleep(0.02)
    rslt = self.i2cbus.read_byte(self._addr)
    return rslt
